Drivers: ADC: Add DMA transfer capability to NXP LPADC

ZhaoxiangJin opened the pull request:
fix https://github.com/zephyrproject-rtos/zephyr/issues/101540
1. added DMA transfer capability to LPADC.
2. added dma-related properties for lpadc node in `nxp_mcxa156.dtsi`.
3. enabled adc_sequence test for frdm_mcxa156, and DMA is used to transfer ADC conversion data in this sample. (In this case, logic channel 0 is connect to the internal VSSA; logic channel 1 is connect to the internal VDDA/4)

<details>
<summary>test logs</summary>

```
*** Booting Zephyr OS build v4.3.0-2968-g622e0a37c192 ***
ADC sequence reading [0]:
- lpadc@400af000, channel 0, 5 sequence samples:
- - 1 = 0mV
- - 1 = 0mV
- - 0 = 0mV
- - 1 = 0mV
- - 1 = 0mV
- lpadc@400af000, channel 1, 5 sequence samples:
- - 982 = 791mV
- - 981 = 790mV
- - 982 = 791mV
- - 982 = 791mV
- - 982 = 791mV
ADC sequence reading [1]:
- lpadc@400af000, channel 0, 5 sequence samples:
- - 1 = 0mV
- - 1 = 0mV
- - 1 = 0mV
- - 1 = 0mV
- - 0 = 0mV
- lpadc@400af000, channel 1, 5 sequence samples:
- - 982 = 791mV
- - 981 = 790mV
- - 982 = 791mV
- - 982 = 791mV
- - 982 = 791mV
ADC sequence reading [2]:
- lpadc@400af000, channel 0, 5 sequence samples:
- - 1 = 0mV
- - 1 = 0mV
- - 1 = 0mV
- - 1 = 0mV
- - 0 = 0mV
- lpadc@400af000, channel 1, 5 sequence samples:
- - 981 = 790mV
- - 981 = 790mV
- - 982 = 791mV
- - 981 = 790mV
- - 982 = 791mV


```

</details>

ZhaoxiangJin committed:
dts: nxp: Add dma properties for nxp mcxa156 lpadc node

Add dmas properties for lpadc node in nxp_mcxa156.dtsi

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
drivers: adc: Adding DMA transfer capability to LPADC

Adding DMA transfer capability to LPADC

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

JarmouniA commented on samples/drivers/adc/adc_sequence/boards/frdm_mcxa156.conf:
> But to be honest, if the ADC sample works normally under the condition of CONFIG_LPADC_ENABLE_DMA=y, then the feature can be considered verified

No not really, that's just a demo, not a real test (with coverage of corner cases & atypical configs ...)

lucien-nxp commented on samples/drivers/adc/adc_sequence/boards/frdm_mcxa156.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright 2025 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_LPADC_ENABLE_DMA=y
```
I think this kconfig don't need to be added manually. 
Please add `depend on` condition like below for LPADC_ENABLE_DMA kconfig 	
`depends on $(dt_compat_any_has_prop,$(DT_COMPAT_NXP_LPC_LPADC),dmas) && $(dt_compat_any_has_prop,$(DT_COMPAT_NXP_LPC_LPADC),dma-names )`

JarmouniA replied:
`default y if ...` is better to leave the option of disabling DMA usage without modifying DT.

JarmouniA replied:
> What is your reason for saying it's not manually set?
I think manual setting is necessary because there are multiple ways for ADC to transfer data - polling? interrupt driven? dma driven?

The assumption is, which is made in many other device drivers too, is that when the user adds `dmas` and `dma-names` properties to the device node, that means they want to use the DMA mode, so there is no need really to burden them with yet another option they have to enable. But it should stay user-selectable in Kconfig.

JarmouniA replied:
> the application layer may still need to use interrupt-based data transfer.

`CONFIG_ADC_LPADC_DMA=n` will do it in that case.

ZhaoxiangJin replied:
To reiterate: DT (Device Tree) describes hardware capabilities, not software policy. The fact that hardware has DMA channels does not mean the application layer should use DMA. Treating `if dmas is written in the device tree, then the user wants to use DMA` as common sense is the root cause of implicit behavior and surprises.

Furthermore, for this kind of unwritten rule, it makes sense to refer to existing implementations. So please take a look at the [STM32 implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/Kconfig.stm32#L29), the [TI implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/Kconfig.cc23x0#L12), and the [SiLabs implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/Kconfig.silabs#L26) (pay attention, SiLabs uses `depends on DMA`, kudos to SiLabs).

JarmouniA commented on drivers/adc/Kconfig.mcux:
```diff
@@ -121,6 +121,14 @@ config LPADC_CHANNEL_COUNT
 
+config LPADC_ENABLE_DMA
+	bool "LPADC DMA support"
+	depends on DMA
```
> In this case, using depends on is more appropriate.

No it's not.

ZhaoxiangJin replied:
The Zephyr [tips ](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/build/kconfig/tips.rst) (`Avoid selecting symbols with prompts or dependencies. Prefer depends on.`) and [Kconfig documentation](https://docs.kernel.org/kbuild/kconfig-language.html#menu-attributes) (`select should be used with care. select will force a symbol to a value without visiting the dependencies.` AND `In general use select only for non-visible symbols (no prompts anywhere) and for symbols with no dependencies.`) clearly record how to properly use select and depends on. If you think they are wrong, please first refute their conclusions.
It's normal that there are many non-standard uses in Zephyr and even Linux, because the world is just a huge mess/shitshow. Using these things as evidence doesn't hold water.

JarmouniA replied:
The Zephyr tips are just that, tips.

zejiang0jason replied:
Hi @JarmouniA 

> It would be hard to come up with a general rule, but maybe a set of recommendations specific to drivers is possible.

That would be helpful, even if just some recommendations for drivers. Clear is better for community driven projects.

ZhaoxiangJin commented on drivers/adc/Kconfig.mcux:
```diff
@@ -121,6 +121,14 @@ config LPADC_CHANNEL_COUNT
 	  15 ADC channels. This value corresponds to how many of the CMD
 	  registers can be configured within the ADC.
 
+config ADC_LPADC_ENABLE_DMA_TRANSFER
```
In my view, there are three categories of naming. The pros and cons I've summarized are as follows:

1. `ADC_LPADC_ENABLE_DMA_TRANSFER`: Most semantically clear ("enable DMA transfer"), but verbose, slightly inconvenient for maintenance and searching; => `Implementation in the current PR`

2. `ADC_<DRIVER>_DMA`: Concise, but semantics not explicit enough (whether it's "enable" or "exists" depends on context). => `Implementation used by stm32, esp32, silabs`

3. `ADC_<DRIVER>_DMA_DRIVEN`: Balances conciseness and semantics (clearly indicates this is "DMA-driven mode"), good readability, but not entirely consistent with existing driver naming conventions. => `TI's implementation`

Personally, I prefer [TI's implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/Kconfig.cc23x0#L12). If you think there's a written regulation that mandates doing it like STM32, then please point it out, and I'm willing to follow (although the semantics aren't clear enough)."

prayK reviewed (approved):
The patch runs great on my end. Thanks to NXP for the quick support.

zejiang0jason commented on samples/drivers/adc/adc_sequence/boards/frdm_mcxa156.conf:
```diff
@@ -2,4 +2,3 @@
 # SPDX-License-Identifier: Apache-2.0
 
 CONFIG_ADC_MCUX_LPADC_DMA_DRIVEN=y
-CONFIG_DMA=y
```
this should be the patch that enabling sample.

zejiang0jason commented on tests/drivers/adc/adc_api/boards/frdm_mcxa156.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright 2025 NXP
```
is it possible to improve testcase.yaml, so that both DMA driven and interrupt driven can be tested?

ZhaoxiangJin replied:
improved the testcase.yaml for nxp lpadc driver by adding dma test scenario

ZhaoxiangJin committed:
samples: adc: enable adc_sequence test for frdm_mcxa156

1. enable adc_sequence test for frdm_mcxa156
2. use DMA to transfer ADC conversion data in this sample

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: drivers: enable adc dma test scenario for frdm_mcxa156

improve test coverage for nxp lpadc driver by adding dma test
scenario.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>
